odoo/orm/models.py

```diff
@@ -1406,7 +1406,10 @@ def search_fetch(
 
         fields_to_fetch = self._determine_fields_to_fetch(field_names)
 
-        return self._fetch_query(query, fields_to_fetch)
+        fetched = self._fetch_query(query, fields_to_fetch)
+        if not self.env.su and self._check_access is BaseModel._check_access:
+            self.env.transaction._add_to_access_cache(fetched)
```

Comment:
Putting the cache on `self.env.transaction` is an implementation detail.  IMO the method should be on `self.env`.